### ISSUE TITLE
Video Tracking with persist=True Exits After First Frame in Frame-by-Frame Processing

### ISSUE THREAD

**Issue by monkeycc** [ultralytics/ultralytics#22614]
# Issue: Video Tracking with `persist=True` Exits After First Frame in Frame-by-Frame Processing

## Description
When using YOLO11 for object tracking with frame-by-frame video reading and setting `persist=True` in each `model.track` call, the program exits after the first frame. If `persist=False` is set, the entire video can be processed normally.

## Steps to Reproduce
https://docs.ultralytics.com/modes/track/#persisting-tracks-loop

```python
import cv2

from ultralytics import YOLO

# Load the YOLO11 model
model = YOLO("yolo11n.pt")

# Open the video file
video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

# Loop through the video frames
while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        # Run YOLO11 tracking on the frame, persisting tracks between frames
        results = model.track(frame, persist=True)

        # Visualize the results on the frame
        annotated_frame = results[0].plot()

        # Display the annotated frame
        cv2.imshow("YOLO11 Tracking", annotated_frame)

        # Break the loop if 'q' is pressed
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        # Break the loop if the end of the video is reached
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()
```

```python
import cv2
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

while cap.isOpened():
    success, frame = cap.read()
    if success:
        results = model.track(frame, persist=True)  # Using persist=True causes only first frame to be processed
        annotated_frame = results[0].plot()
        cv2.imshow("YOLO11 Tracking", annotated_frame)
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

cap.release()
cv2.destroyAllWindows()
```

## Expected Behavior
Even when using frame-by-frame video reading with `persist=True`, it should be able to track the entire video normally while maintaining tracking ID continuity.

## Actual Behavior
The program only displays the first frame and then exits the loop.

## Additional Information
Through debugging, it was found that when using `persist=True`, during the second frame reading, `cap.read()` returns `success` as `False`, causing the loop to exit. However, if `persist=True` is not used, the video can be read normally. The video file itself is normal.

## Suggested Fixes
Possible reasons include:
1. When using `persist=True`, the tracker state may be causing issues in frame-by-frame processing, leading to abnormal video reading.
2. Alternatively, `persist=True` should not be used in frame-by-frame processing because the tracker should automatically maintain state throughout the video sequence without explicitly setting `persist`.

According to the documentation, the `persist` parameter is used to maintain trackers between different sequences. Within the same video sequence, the tracker should automatically maintain state without setting `persist`. Therefore, when processing the same video frame-by-frame, `persist=False` (default) should be used or the parameter should be omitted.

However, we noticed that in official examples, `persist=True` is sometimes used to maintain trackers across multiple videos. So this parameter might be misunderstood.

We recommend:
1. Clearly documenting the usage of the `persist` parameter in the documentation
2. Providing examples of how to correctly use tracking in frame-by-frame processing
3. If using `persist=True` in frame-by-frame processing causes video reading interruption, this might be a bug that needs fixing

Additionally, it would be helpful to:
- Add parameter validation to warn users when `persist=True` is used in inappropriate scenarios
- Provide better error messages to guide users toward correct usage
- Consider supporting `persist=True` in frame-by-frame scenarios if technically feasible

**Comment by UltralyticsAssistant** [comment]
👋 Hello @monkeycc, thanks for opening this issue and for the clear description and code snippet 🚀 This is an automated response to help get things moving quickly. An Ultralytics engineer will review and assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

To accelerate triage for your tracking/persist report, please include:
-`ultralytics` version (e.g., output of `python -c "import ultralytics, torch, sys; print(ultralytics.__version__, torch.__version__, sys.version)"`)
-OS, Python, PyTorch, CUDA versions, GPU model
-Exact model and assets used (e.g., `yolo11n.pt`) and any custom tracker config if applicable
-A minimal script that reproduces the issue (you’ve shared one—thank you). If possible, attach or describe a very short video that triggers the behavior
-Full console logs including any warnings when `persist=True` vs. `persist=False`
-Confirmation whether the behavior occurs with a different short MP4 and with CPU vs. GPU runs

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the detailed report. Once you share the MRE details above, we’ll investigate the `persist=True` frame-by-frame tracking behavior with YOLO11 and follow up promptly 🔍💡

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by monkeycc** [comment]
`pip uninstall opencv-python opencv-contrib-python opencv-python-headless opencv-contrib-python-headless -y`
`pip install opencv-contrib-python`

I feel like it's an issue with OpenCV
I have reinstalled and the issue has been resolved

**Comment by glenn-jocher** [comment]
Great to hear a clean OpenCV reinstall resolved it. For frame-by-frame loops your `persist=True` usage is correct—it only keeps tracker state across calls and doesn’t touch `cv2.VideoCapture` (see the [model.track reference](https://docs.ultralytics.com/reference/engine/model/#track)). To avoid similar issues, don’t mix headless and non‑headless OpenCV builds; if it recurs, please attach the full traceback and `yolo checks` output. We’ll tighten the `persist` wording in the docs, and PRs to clarify usage are welcome.

**Comment by ZakQuddoos** [comment]
Hi there, I am having this same issue. I am running the `model.track()` method on individual frames using the code provided in the documentation. Per this discussion, I have uninstalled OpenCV and reinstalled the latest version. This has not fixed the issue, here is the output when I run yolo checks. 

```
Ultralytics 8.3.232  Python-3.12.10 torch-2.6.0+cu124 CUDA:0 (NVIDIA RTX 3500 Ada Generation Laptop GPU, 12282MiB)
Setup complete  (32 CPUs, 31.7 GB RAM, 909.9/951.6 GB disk)

OS                     Windows-11-10.0.26200-SP0
Environment            Windows
Python                 3.12.10
Install                git
Path                   C:\Users\ZakQuddoos\OneDrive - Bytronic\Documents\MagnaPower\Inspect\inspect\inspect_env\Lib\site-packages\ultralytics
RAM                    31.68 GB
Disk                   909.9/951.6 GB
CPU                    13th Gen Intel Core(TM) i9-13980HX
CPU count              32
GPU                    NVIDIA RTX 3500 Ada Generation Laptop GPU, 12282MiB
GPU count              1
CUDA                   12.4

numpy                   2.2.6>=1.23.0
matplotlib              3.10.7>=3.3.0
opencv-python           4.12.0.88>=4.6.0
pillow                  11.3.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.3>=1.4.1
torch                   2.6.0+cu124>=1.8.0
torch                   2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.21.0+cu124>=0.9.0
psutil                  7.1.3>=5.8.0
polars                  1.35.2>=0.20.0
ultralytics-thop        2.0.18>=2.0.18
```

**Comment by monkeycc** [comment]
You can try an older version of OpenCV

I don't think the latest OpenCV version is supported here
Or where there is a conflict

**Comment by glenn-jocher** [comment]
Thanks for the details @ZakQuddoos and for the follow‑up @monkeycc.  

From your `yolo checks` output the Ultralytics install looks fine, and since the loop exits without an exception the only thing that can break the `while cap.isOpened()` loop is `cap.read()` returning `success=False` before `model.track()` is called, which points to OpenCV/codec behavior rather than the tracker itself (the tracker only sees the `numpy` frame; you can see how `persist` is wired in the [tracker callbacks reference](https://docs.ultralytics.com/reference/trackers/track/)).  

To narrow this down, could you please:  
1) Run a pure OpenCV loop with the same video and confirm whether it also stops after the first frame:  
```python
import cv2

video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

while cap.isOpened():
    success, frame = cap.read()
    if not success:
        print("read() returned False")
        break
    cv2.imshow("Test", frame)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

cap.release()
cv2.destroyAllWindows()
```  
2) If that works, try driving the tracker on synthetic frames (to decouple from `VideoCapture`):  
```python
import numpy as np
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
dummy = np.zeros((480, 640, 3), dtype=np.uint8)

for i in range(100):
    results = model.track(dummy, persist=True)
```
and let us know if either of these reproduces the early exit.  

If the pure OpenCV loop already fails, it’s very likely a regression or codec issue in 4.12.0.88, so trying a 4.10.x build as @monkeycc suggested is a good next step; if it only fails once `model.track()` is in the loop, please share the exact minimal script and any console output so we can try to reproduce on our side.

**Comment by ZakQuddoos** [comment]
Hi Glen,

Thank you for the help, I installed an older version of OpenCV and it worked.

**Comment by glenn-jocher** [comment]
Good to hear the downgrade fixed it; I’d recommend pinning that `opencv-python` version in your environment to avoid surprise regressions, and if you don’t mind, please drop the exact working version here so others running into the same issue can benefit.
